### DISCUSSION TITLE
Delete Templates from specific secondary storages

### DISCUSSION THREAD

**Discussion by nxsbi** [apache/cloudstack#9229] [Ideas]
##### ISSUE TYPE

 * Other

##### COMPONENT NAME

~~~
UI
~~~

##### CLOUDSTACK VERSION

~~~
4.19
~~~

##### SUMMARY

Unable to delete Template from specific secondary storage from UI. 

##### STEPS TO REPRODUCE

~~~

I have 2 secondary storage. One is to be retired. I have already moved all snapshots to the other one. However, I am not able to delete secondary storage because several Templates have been downloaded to both Sec Store. UI does not have option to delete Template from specific secondary store, it only deleted from the entire Zone. 

How to I do this - is there cloudmonkey option? Or manipulate database? 

~~~

##### EXPECTED RESULTS

~~~
Able to delete template for specific Secondary Storage. 
~~~

##### ACTUAL RESULTS

~~~
Not able to delete template from specific secondary storage. 
~~~

**Comment by DaanHoogland** [comment]
@nxsbi , you are right. This would be a good addition. As a workaround you can look at template_store_ref. It does not have a removed field but for your use case it should, and next some API functionality should be added. 
Please let us know how you fare.

**Comment by Pearl1594** [comment]
@nxsbi Maybe you could try the `migrateResourceToAnotherSecondaryStorage` api / operation to move the templates from the store you want to remove? Not really sure how this would behave with public template, but worth a shot.

**Comment by nxsbi** [comment]
For anyone else trying to do this - As of 4.19.0.1, if you want to delete the secondary storage (assuming you have 2 or more secondary storage) that has templates stored on it, I followed the below steps. Note that you would want to move all your snapshots to other secondary storage beforehand. 

1. Identify the ID for the image store to be deleted. I simply ran the query below and manually noted it down. 

`select * from image_store where name = 'name of secondary store to be deleted';`

2. Verify that all templates are available on the other secondary storage. - NOTE I did not have any missing ones so did not have to address this issue, but if you have any missing ones, you will need to explore how to copy them over to the other image store. 

```
-- Identify templates that only exist in the secondary store to be deleted

SELECT DISTINCT tsr1.template_id
FROM template_store_ref  tsr1
WHERE tsr1.download_state = 'DOWNLOADED'
  AND tsr1.state = 'Ready'
  AND tsr1.store_id = <your_store_id_to_be_deleted>
  AND NOT EXISTS (
    SELECT 1
    FROM template_store_ref  tsr2
    WHERE tsr2.download_state = 'DOWNLOADED'
      AND tsr2.state = 'Ready'
      AND tsr2.store_id in (<your_store_id_that_will_not_be_deleted>) 
      AND tsr2.template_id = tsr1.template_id
  );
```

3. Delete the records from template_store_ref

```
-- Take a backup
create table template_store_ref_backup as select * from template_store_ref;
-- Delete the records
delete from template_store_ref where store_id = <your_store_id_to_be_deleted>;

```
4. Go to the UI, and set the Secondary Store to read only, 

5. From UI, Delete the secondary Store

Verify everything is working on the UI as expected.
